### PR TITLE
Default to using mask mode in benchmarks

### PR DESCRIPTION
**Pull request by magnumripper**

### CODE REVIEW

File: doc/OPTIONS [hunk 1]
```diff
@@ -194,13 +194,20 @@ file will be used as the format type.
...
 --stress-test[=TIME]		continuous self-test
 
 Perform self-tests just like with --test except it loops until failure or
 until aborted by user.  This is mostly for debugging.
 
+--no-mask			benchmark using regular test vectors
+
+This is used together with --test.  By default the benchmark is made using
+mask mode, which with some OpenCL or ZTEX formats is accelerated fully or
+partially onm device-side.  Adding this option switches back to the legacy
```
Comment (magnumripper) [inline doc/OPTIONS]: s/onm/on

**Comment by magnumripper** [comment]
@solardiz we've got a problem here. While the bot went tits up under Wine, the problem is also seen in Linux ASan builds using `--test-fuill=0` (not sure why build bots doesn't seem to catch that).

I believe all Jumbo formats can reset/deallocate/destroy fully at `done()` and then be re-used with `init()`. Most core formats can't. I'm not quite sure what happens here but it's likely something with bsdicrypt's `init()` calling `DES_*_init()` which isn't fully reset after DEScrypt.

I don't have time for this right now.

**Comment by magnumripper** [comment]
Note that the problem was introduced with 7187208db, already merged. This commit just make it happen by default...

**Comment by solardiz** [comment]
The circleci report here says "now testing the 32-bit SSE2 Windows build", but then somehow actually tests a SIMD-less 32-bit build (a bug in our CI setup?), which crashes on hmac-md5. I guess that's a different issue from what you're referring to, @magnumripper? Yet you write "While the bot went tits up under Wine, the problem is also seen ..." suggesting you think it's the same problem?

**Comment by magnumripper** [comment]
Not sure what bot I saw but that was BSDI

Here's what I see on OSX
```
$ make -sj8 && ../run/john --test-full=0 -mask

Make process completed.
Will run 8 OpenMP threads
Testing: descrypt, traditional crypt(3) [DES 256/256 AVX2]... (8xOMP) PASS 
Testing: bsdicrypt, BSDI crypt(3) ("_J9..", 725 iterations) [DES 256/256 AVX2]... (8xOMP) PASS =================================================================
==64650==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x619000000980 at pc 0x00010566b7fc bp 0x7ffeea5c8540 sp 0x7ffeea5c8538
READ of size 1 at 0x619000000980 thread T0
    #0 0x10566b7fb in set_key BSDI_fmt.c:300
    #1 0x105b46ed6 in bench_generate_keys mask.c:1746
    #2 0x105b4d8db in do_mask_crack mask.c:2599
    #3 0x105af7fa1 in bench_set_keys bench.c:275
    #4 0x105af8e0f in benchmark_format bench.c:458
    #5 0x105afa582 in benchmark_all bench.c:796
    #6 0x105f5d1c2 in main john.c:1700
    #7 0x7fff650d63d4 in start (libdyld.dylib:x86_64+0x163d4)

0x619000000980 is located 0 bytes to the right of 1024-byte region [0x619000000580,0x619000000980)
allocated by thread T0 here:
    #0 0x107effcdd in wrap_malloc (libasan.5.dylib:x86_64+0x6bcdd)
    #1 0x105b4ed3e in mem_alloc memory.c:76
    #2 0x105b4bea3 in mask_init mask.c:2201
    #3 0x105afaad0 in benchmark_all bench.c:791
    #4 0x105f5d1c2 in main john.c:1700
    #5 0x7fff650d63d4 in start (libdyld.dylib:x86_64+0x163d4)

SUMMARY: AddressSanitizer: heap-buffer-overflow BSDI_fmt.c:300 in set_key
Shadow bytes around the buggy address:
  0x1c32000000e0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c32000000f0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c3200000100: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c3200000110: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c3200000120: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x1c3200000130:[fa]fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c3200000140: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x1c3200000150: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c3200000160: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c3200000170: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x1c3200000180: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==64650==ABORTING
Abort trap: 6
```

I'm seeing other problems as well (with Jumbo formats too). Need to test more.

**Comment by magnumripper** [comment]
Hold on... some of this might be caused by me doing a PR for each separate fix. I now rebased upon bleeding (with other changes merged) and force pushed. Let's see how that goes.

**Comment by magnumripper** [comment]
This is confusing. The failing tests are using `-test-full=0` so they shouldn't even reset the format (since they won't benchmark anyway).

**Comment by claudioandre-br** [comment]
It is not easy to make it fail. This is `crypt_all()`.
```
Backtrace:
=>0 0x0000000000517787 crypt_all+0x1d7(pcount=<internal error>, salt=<internal error>) [/home/JtR/project/src/HDAA_fmt_plug.c:491] in john (0x000000000033f6a0)
  1 0x00000000006571b4 in john (+0x2571b3) (0x000000000033f740)
  2 0x00000000006578c0 benchmark_all+0x30f() [/home/JtR/project/src/bench.c:796] in john (0x000000000033f930)
  3 0x000000000066dbfa main+0x1b7f(argc=<internal error>, argv=<internal error>) [/home/JtR/project/src/john.c:1700] in john (0x000000000033fac0)
  4 0x00000000004013f7 in john (+0x13f6) (0x0000000002a724c0)
```

https://circleci.com/gh/claudioandre-br/JohnTheRipper/1919

**Comment by claudioandre-br** [comment]
>  problem is also seen in Linux ASan builds using --test-fuill=0 (not sure why build bots doesn't seem to catch that).

They do. But only clang.

```
Testing: bsdicrypt, BSDI crypt(3) ("_J9..", 725 iterations) [DES 128/128 AVX]... (2xOMP) DES_bs_b.c:845:33: runtime error: addition of unsigned offset to 0x7f13e1d89cc0 overflowed to 0x7f13e1d84000
SUMMARY: UndefinedBehaviorSanitizer: undefined-behavior DES_bs_b.c:845:33 in 
PASS =================================================================
==9878==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x619000000e80 at pc 0x0000005c66b1 bp 0x7fff7dfac2d0 sp 0x7fff7dfac2c8
READ of size 1 at 0x619000000e80 thread T0
    #0 0x5c66b0 in set_key /cwd/src/BSDI_fmt.c:300:7
    #1 0xf7c4f0 in bench_generate_keys /cwd/src/mask.c:1746:8
    #2 0xf77dd2 in do_mask_crack /cwd/src/mask.c:2599:8
    #3 0xeea235 in bench_set_keys /cwd/src/bench.c:275:8
    #4 0xee8248 in benchmark_format /cwd/src/bench.c:458:2
    #5 0xeec008 in benchmark_all /cwd/src/bench.c:796:17
    #6 0xf3f8e6 in john_run /cwd/src/john.c:1700:17
    #7 0xf3e4c0 in main /cwd/src/john.c:2115:2
    #8 0x7f13e5444b6a in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x26b6a)
    #9 0x41dd49 in _start (/cwd/run/john+0x41dd49)
0x619000000e80 is located 0 bytes to the right of 1024-byte region [0x619000000a80,0x619000000e80)
allocated by thread T0 here:
    #0 0x4c96f3 in __interceptor_malloc /build/llvm-toolchain-8-F3l7P1/llvm-toolchain-8-8/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:146:3
    #1 0xf94263 in mem_alloc /cwd/src/memory.c:76:14
    #2 0xf706d6 in mask_init /cwd/src/mask.c:2201:17
    #3 0xeebf56 in benchmark_all /cwd/src/bench.c:791:4
    #4 0xf3f8e6 in john_run /cwd/src/john.c:1700:17
    #5 0xf3e4c0 in main /cwd/src/john.c:2115:2
    #6 0x7f13e5444b6a in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x26b6a)
SUMMARY: AddressSanitizer: heap-buffer-overflow /cwd/src/BSDI_fmt.c:300:7 in set_key
Shadow bytes around the buggy address:
  0x0c327fff8180: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c327fff8190: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c327fff81a0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c327fff81b0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c327fff81c0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0c327fff81d0:[fa]fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c327fff81e0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c327fff81f0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c327fff8200: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c327fff8210: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c327fff8220: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
  Shadow gap:              cc
==9878==ABORTING
The command ".travis/travis-ci.sh" exited with 1.
```

**Comment by claudioandre-br** [comment]
Note: ASAN + UBSAN above (BTW: UBSAN is complaining).

ASAN only
```
Testing: bsdicrypt, BSDI crypt(3) ("_J9..", 725 iterations) [DES 128/128 SSE2]... PASS =================================================================
==13678==ERROR: AddressSanitizer: heap-buffer-overflow on address 0x619000000e80 at pc 0x0000004f7ff3 bp 0x7fffb69b7e90 sp 0x7fffb69b7e88
READ of size 1 at 0x619000000e80 thread T0
    #0 0x4f7ff2 in set_key /home/travis/build/claudioandre-br/JohnTheRipper/src/BSDI_fmt.c:300:7
    #1 0x9d08b0 in bench_generate_keys /home/travis/build/claudioandre-br/JohnTheRipper/src/mask.c:1746:8
    #2 0x9ce8f1 in do_mask_crack /home/travis/build/claudioandre-br/JohnTheRipper/src/mask.c:2599:8
    #3 0x987840 in bench_set_keys /home/travis/build/claudioandre-br/JohnTheRipper/src/bench.c:275:8
    #4 0x986921 in benchmark_format /home/travis/build/claudioandre-br/JohnTheRipper/src/bench.c:458:2
    #5 0x988545 in benchmark_all /home/travis/build/claudioandre-br/JohnTheRipper/src/bench.c:796:17
    #6 0x9b1191 in john_run /home/travis/build/claudioandre-br/JohnTheRipper/src/john.c:1700:17
    #7 0x9b053d in main /home/travis/build/claudioandre-br/JohnTheRipper/src/john.c:2115:2
    #8 0x7f8c61bdaf44 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21f44)
    #9 0x41c79b in _start (/home/travis/build/claudioandre-br/JohnTheRipper/run/john+0x41c79b)
0x619000000e80 is located 0 bytes to the right of 1024-byte region [0x619000000a80,0x619000000e80)
allocated by thread T0 here:
    #0 0x4bd973 in __interceptor_malloc /local/mnt/workspace/tmp/ubuntu_rel/llvm/utils/release/final/llvm.src/projects/compiler-rt/lib/asan/asan_malloc_linux.cc:67:3
    #1 0x9f8bf6 in mem_alloc /home/travis/build/claudioandre-br/JohnTheRipper/src/memory.c:76:14
    #2 0x9c9c20 in mask_init /home/travis/build/claudioandre-br/JohnTheRipper/src/mask.c:2201:17
    #3 0x9884d8 in benchmark_all /home/travis/build/claudioandre-br/JohnTheRipper/src/bench.c:791:4
    #4 0x9b1191 in john_run /home/travis/build/claudioandre-br/JohnTheRipper/src/john.c:1700:17
    #5 0x9b053d in main /home/travis/build/claudioandre-br/JohnTheRipper/src/john.c:2115:2
    #6 0x7f8c61bdaf44 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x21f44)
SUMMARY: AddressSanitizer: heap-buffer-overflow /home/travis/build/claudioandre-br/JohnTheRipper/src/BSDI_fmt.c:300:7 in set_key
Shadow bytes around the buggy address:
  0x0c327fff8180: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c327fff8190: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c327fff81a0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c327fff81b0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x0c327fff81c0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
=>0x0c327fff81d0:[fa]fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c327fff81e0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c327fff81f0: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c327fff8200: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c327fff8210: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
  0x0c327fff8220: fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa fa
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==13678==ABORTING
The command ".travis/travis-ci.sh" exited with 1.
```

**Comment by claudioandre-br** [comment]
> This is confusing. The failing tests are using -test-full=0 so they shouldn't even call formats' done/reset (since they won't benchmark anyway).

A `-test-full=0` run is benchmarking. Note the **PASS** plus autotune below.

```
../run/john --test-full=0 --format=sha512crypt-opencl -dev=2 -v:5
Device 2: Juniper
Testing: sha512crypt-opencl, crypt(3) $6$ (rounds=5000) [SHA512 OpenCL]... Loaded 6 hashes with 4 different salts to test db from test vectors
[...]
Local worksize (LWS) 7, global worksize (GWS) 49 (7 blocks)
PASS
Calculating best GWS for LWS=64; max. 100ms single kernel invocation.
Raw speed figures including buffer transfers:
xfer: 26.445us, crypt: 119x2.596ms, xfer back: 41.222us, prep: 5.365ms, pp: 101.222us, final: 141.111us, var: 2.585ms/2.578ms
gws:      1024	   3265c/s    16325000 rounds/s 313.611ms per crypt_all()!
xfer: 50.333us, crypt: 119x2.612ms, xfer back: 80.444us, prep: 6.184ms, pp: 131.111us, final: 149.445us, var: 2.606ms/2.608ms
```

**Pull request closed without merging**